src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs

```diff
@@ -0,0 +1,321 @@
+                    }
+                }
+                else
+                {
+                    context.ReportDiagnostic(symbol.CreateDiagnostic(GeneralPreviewFeatureAttributeRule, symbol.Name));
+                }
+            }
+        }
+
+        private static void BuildSymbolInformationFromOperations(OperationAnalysisContext context, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols)
+        {
+            IOperation operation = context.Operation;
+            if (OperationUsesPreviewFeatures(operation, requiresPreviewFeaturesSymbols, out ISymbol? symbol))
+            {
+                if (operation is ICatchClauseOperation catchClauseOperation)
+                {
+                    operation = catchClauseOperation.ExceptionDeclarationOrExpression;
+                }
+
+                context.ReportDiagnostic(operation.CreateDiagnostic(GeneralPreviewFeatureAttributeRule, symbol.Name));
+            }
+        }
+
+        private static bool OperationUsesPreviewFeatures(IOperation operation, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols, [NotNullWhen(true)] out ISymbol? symbol)
+        {
+            symbol = GetOperationSymbol(operation);
+            if (symbol == null)
+            {
+                return false;
+            }
+
+            return OperationUsesPreviewFeatures(symbol, requiresPreviewFeaturesSymbols);
+        }
+
+        private static ISymbol? GetOperationSymbol(IOperation operation)
+            => operation switch
+            {
+                IInvocationOperation iOperation => iOperation.TargetMethod,
+                IObjectCreationOperation cOperation => cOperation.Constructor,
+                IPropertyReferenceOperation pOperation => pOperation.Property,
+                IFieldReferenceOperation fOperation => fOperation.Field,
+                IDelegateCreationOperation dOperation => dOperation.Type,
+                IEventReferenceOperation eOperation => eOperation.Member,
+                IUnaryOperation uOperation => uOperation.OperatorMethod,
+                IBinaryOperation bOperation => bOperation.OperatorMethod,
+                IArrayCreationOperation arrayCreationOperation => (arrayCreationOperation.Type as IArrayTypeSymbol)?.ElementType,
+                ICatchClauseOperation catchClauseOperation => catchClauseOperation.ExceptionType,
+                _ => null,
+            };
+
+        private static bool OperationUsesPreviewFeatures(ISymbol symbol, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols)
+        {
+            var containingSymbol = symbol.ContainingSymbol;
+            // Namespaces do not have attributes
+            while (containingSymbol is INamespaceSymbol)
+            {
+                containingSymbol = containingSymbol.ContainingSymbol;
+            }
+
+            if (containingSymbol != null)
+            {
+                if (HasPreviewAttribute(containingSymbol, requiresPreviewFeaturesSymbols))
```

Comment:
Please move the logic to the `ContainingTypeIsPreview` method  https://github.com/dotnet/roslyn-analyzers/blob/f147d906818271152d87beac60a6ab3c6808772b/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs#L138 and use the method here

Reply:
That method doesn't special case INamespaceSymbol though. I can do it if you think it's important, but I'd prefer to keep it this way.

Reply:
> That method doesn't special case INamespaceSymbol though.

That is why I said move this logic, i do not understand why one of the same logic special case it but the other not, if you have a reason for that then its OK, you can keep it as is

Reply:
As it turns out the other method will never see a NamespaceSymbol, so it was safe to move this logic there.